Availables pg versions

fluca1978 opened the pull request:
Implements the `availables`  command that shows the downloadable versions.

theory commented:
Oh, nice. Might I suggest we use the term `available` instead of `availables`? The plural form sounds weird.

xzilla commented:
I'd also suggest using sort -n, which will put them in release order rather than alphanumeric order.

fluca1978 commented:
I've made a refactoring of this command.
Since I was unable to sort decently the output (and I did not want to introduce a Perl depency, but with an hash and map it would have been a metter of ltwo lines), I've created a couple of cycles to sort all versions by brand/major version number.

Now the command displays an output similar to the following:
```
PostgreSQL 11 available versions
==============================================
11beta1 11beta2 11beta3
==============================================
PostgreSQL 10 available versions
==============================================
10.0    10.1    10.2    10.3    10.4    10.5
==============================================
```

theory commented on bin/pgenv:
```diff
@@ -275,6 +276,46 @@ EOF
+        done
+
+        # extra remove of abandoned files (should never happen)
+        rm /tmp/pgenv.txt.* > /dev/null 2>&1
```
Should se a trap function, instead, so if there's a premature exit, the temp files still get cleaned up.

theory commented on bin/pgenv:
```diff
@@ -275,6 +276,46 @@ EOF
+        # First of all download the page with the links to each version (href=va.b.c/)
+        # and trim out each version, then extract the major version and produce
+        # a text file with all versions under the brand/major version
+        for version in $(curl http://ftp.postgresql.org/pub/source/ | grep -o '<a href=['"'"'"][^"'"'"']*['"'"'"]' | sed -e 's/^<a href=["'"'"']//' -e 's/["'"'"']$//' | sed 's/^v//' | sed 's/\/$//')
```
Oy, I don't suppose anyone knows of a plain-text listing of all versions…

theory commented:
Hrm…

```
 > pgenv available
	Available PostgreSQL versions
	=============================
ls: /tmp/pgenv.txt.*: No such file or directory
```

:-(

fluca1978 commented:
I cannot reproduce the error on my system right now, I've added an error message to prevent the weird error you got, in the hope it can help understanding what is going wrong. Should not be a curl failure, since there's a trap, but seems no useful version list is found at all.

fluca1978 commented:
Also I've fixed the command output to reflect the PostgreSQL major version numbering, such as:

```
==============================================
   PostgreSQL major version 10  
==============================================
10.0    10.1    10.2    10.3    10.4    10.5
==============================================

==============================================
   PostgreSQL major version 9.6  
==============================================
9.6.0   9.6.1   9.6.2   9.6.3   9.6.4   9.6.5
9.6.6   9.6.7   9.6.8   9.6.9   9.6.10
==============================================

==============================================
   PostgreSQL major version 9.5  
==============================================
9.5.0   9.5.1   9.5.2   9.5.3   9.5.4   9.5.5
9.5.6   9.5.7   9.5.8   9.5.9   9.5.10  9.5.11
9.5.12  9.5.13  9.5.14
==============================================
```

fluca1978 commented:
I still cannot reproduce the failure in `ls`, but thinking about it could have been something wrong with `/tmp` reference? Therefore I placed now temporary files into a temporary directory created with `mktemp`, this should solve the problem at all.

theory commented:
```
 > pgenv available
mktemp: illegal option -- p
usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
       mktemp [-d] [-q] [-u] -t prefix 
```

I'm on macOS. 

Maybe it'd be better to append to variables rather than use temporary files? Looks like [one can use +=](https://stackoverflow.com/questions/2250131/how-do-you-append-to-an-already-existing-string#2250199) to append to variables in bash.

theory commented:
BTW, I'm going on vacation for a few days, so probably won't be able to look at this again before Monday. Sorry for the hassle, really appreciate your dilligence on this, @fluca1978!

fluca1978 commented:
Not so readable as I'd like, but now it uses only hashes with a little trick to keep versions sorted. However the output has reversed its listing (I've added it in the documentation). I've removed any reference to the temporary directory.

fluca1978 commented:
I suspect there is some kind of problem with bash hashing (that would have been much easier in Perl now!). The sed-grep part did not changed.
Could you please try the last checkout with PGENV_DEBUG set to a value to see where it is getting confused? It should print a lot of messages like:

```
$ PGENV_DEBUG=true pgenv available

        Available PostgreSQL versions
        =============================
[DEBUG] Extracted major version [..]
[DEBUG] Extracted major version [1]
[DEBUG] Normalized major version [10]
[DEBUG] Extracted major version [1]
[DEBUG] Normalized major version [10]
[DEBUG] Extracted major version [10]

```

and should try to print the hash content in the case there are no keys (hoping something can lead to a solution).
Also knowing the bash version could provide some hint:

```
% bash --version
GNU bash, version 4.4.7(1)-release (x86_64-pc-linux-gnu)
```

fluca1978 commented:
BTW, I don't think parsing the git tag list would be easier, since it requires still editing with sed to remove REL, transformign the underscores into dots, and so on.

fluca1978 commented:
Should have fixed with commit d6d29f3.
Now, while it is true that Perl is listed as a dependency, it has been used only to compile it. I'm not sure we should use Perl as a dependency for run-time stuff, since my hope is to make that `--with-perl` configurable in the future (so to remove the dependency). However, I agree that in case this does not work we should implement `available` using Perl.

theory commented:
Okay, I squashed your changes and made some editorial changes. See #8. If it looks good to you, @fluca1978, let's merge it!
